Cleanup created accounts and buckets after tests

williamlardier opened the pull request:
Without cleanup, the GetRolesForWebIdentity API ends up being very slow, eventually causing Vault to be unresponsive and thus causing timeouts, and tests failures. This API is not supposed to work with more than 100 accounts (as per requirements), and here, we end up with >1000, meaning at least this number of internal (and sequential) DB calls per "GetRolesForWebIdentity" API call.

The logic detects that we created an account (during cross-account tests), and delete all resources, before deleting the account.

Issue: ZENKO-4898

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/zenko/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4898 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.10.0`

* `2.6.68`

* `2.7.64`

* `2.8.44`

* `2.9.20`


Please check the `Fix Version/s` of ZENKO-4898, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4898 contains:


* `2.2.10`

* `2.6.68`

* `2.7.64`

* `2.8.44`

* `2.9.20`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.10.0`

* `2.6.68`

* `2.7.64`

* `2.8.44`

* `2.9.20`


Please check the `Fix Version/s` of ZENKO-4898, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4898 contains:


* `2.10.0`

* `2.6.68`

* `2.7.64`

* `2.8.44`

* `2.9.20`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.10.0`

* `2.6.68`

* `2.7.64`

* `2.8.44`

* `2.9.21`


Please check the `Fix Version/s` of ZENKO-4898, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

ghost commented:
/create_integration_branches

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.7/bugfix/ZENKO-4898` with contents from `bugfix/ZENKO-4898`
and `development/2.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.7/bugfix/ZENKO-4898 origin/development/2.7
 $ git merge origin/bugfix/ZENKO-4898
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/bugfix/ZENKO-4898
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-4898` into
`development/2.6`
* [w/2.7/bugfix/ZENKO-4898](https://github.com/scality/zenko/compare/development/2.7...w/2.7/bugfix/ZENKO-4898) will be merged into `development/2.7` 
* [w/2.8/bugfix/ZENKO-4898](https://github.com/scality/zenko/compare/development/2.8...w/2.8/bugfix/ZENKO-4898) will be merged into `development/2.8` 
* [w/2.9/bugfix/ZENKO-4898](https://github.com/scality/zenko/compare/development/2.9...w/2.9/bugfix/ZENKO-4898) will be merged into `development/2.9` 
* [w/2.10/bugfix/ZENKO-4898](https://github.com/scality/zenko/compare/development/2.10...w/2.10/bugfix/ZENKO-4898) will be merged into `development/2.10` 



The following branches will **NOT** be impacted:

* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

williamlardier commented on .github/actions/archive-artifacts/action.yaml:
```diff
@@ -42,3 +42,22 @@ runs:
+
+        MONGODB_COLLECTIONS=$(kubectl exec -n ${NAMESPACE} data-db-mongodb-sharded-mongos-0 -- mongo ${ZENKO_MONGODB_DATABASE} -u ${MONGODB_ROOT_USERNAME} -p ${MONGODB_ROOT_PASSWORD} --authenticationDatabase admin --quiet --eval "db.getCollectionNames().join(' ')")
+
+        for collection in ${MONGODB_COLLECTIONS}; do
```
do we want to keep only the special collections?
For now, I see ~400 collections kept (due to object lock or test failing), they all have less than 10 objects anyway

francoisferrand replied:
not sure what we want to do...

it takes ~1min20, which is significant (maybe not after 2h of test, but we need to reduce this 2h anyway...)

- we should try to reduce the time. Probably a large time is related to kubectl kube setup & I/O transfers out of the container... May be much faster using mongo dump ; and possibly "copying" the file from the container after the tool has finished, or creating a new container with a host-volume to directly "save" the file to the host...
- if we can't make it significantly faster, we may need some filtering indeed AND/OR do this only if the previous job has failed...

francoisferrand commented on tests/ctst/common/hooks.ts:
```diff
@@ -37,8 +49,129 @@ Before({ tags: '@Quotas', timeout: 1200000 }, async function (scenarioOptions) {
 
+After(async function (this: Zenko, results) {
+    if (results.result?.status === 'FAILED') {
+        return;
```
can we add a log here, so it is easy to see that something was not cleand up, and to make an explicit link between a specific test and what bucket/account/... were used (and retained in DB)

francoisferrand commented on tests/ctst/common/hooks.ts:
```diff
@@ -37,8 +49,129 @@ Before({ tags: '@Quotas', timeout: 1200000 }, async function (scenarioOptions) {
     await teardownQuotaScenarios(this as Zenko);
 });
 
+After(async function (this: Zenko, results) {
```
this seems mostly related to the iam tests: and we actually don't do anything if there is no saved `crossAccountName`...
- similar to `teardownQuotaScenarios`, would be better to move this cleanup logic to the associated test suite (even if we need to keep the `After` here) : to have better "locality" of the test & associated cleanup code
- since it is specific to these tests, maybe we should use a "tag filter" (like what is done for quotas) to better express this should happen only for these specific tests? (instead of just relying on `crossAccountName`). It would be more explicit.

williamlardier replied:
-> ZENKO-4904

francoisferrand commented on tests/ctst/world/Zenko.ts:
```diff
@@ -943,6 +951,16 @@ export default class Zenko extends World<ZenkoWorldParameters> {
+        this.addToSaved('createdObjects', createdObjects);
+    }
+
+    getCreatedObjects() {
```
it seems often (or always?) use this to then get an object: `this.getCreatedObjects()?.get(objName)`
so we probably should have a `getCreatedObject(objectName string)` function instead/in addition.

francoisferrand commented on tests/ctst/steps/utils/utils.ts:
```diff
@@ -197,17 +197,18 @@ async function createBucketWithConfiguration(
+    const versionId = extractPropertyFromResults<string>(result, 'VersionId');
+    world.saveCreatedObject(world.getSaved<string>('objectName'), versionId || '');
+
     world.addToSaved('versionId', extractPropertyFromResults(
```
saving `versionId` seems redundant with `world.saveCreatedObject` ; even if we keep it to avoid too wide-ranging changes to the tests, should it not be done inside `saveCreatedObject` ?

williamlardier replied:
I removed the lines, and updated the function: it was storing always the last version id, but we may need to store all the known version IDs...

williamlardier replied:
-> ZENKO-4904

francoisferrand commented on tests/ctst/steps/utils/utils.ts:
```diff
@@ -197,17 +197,18 @@ async function createBucketWithConfiguration(
     world.addToSaved('objectName', objectName || Utils.randomString());
-    const objectNameArray = world.getSaved<string[]>('objectNameArray') || [];
-    objectNameArray.push(world.getSaved<string>('objectName'));
-    world.addToSaved('objectNameArray', objectNameArray);
     await uploadSetup(world, 'PutObject');
     world.addCommandParameter({ key: world.getSaved<string>('objectName') });
     world.addCommandParameter({ bucket: world.getSaved<string>('bucketName') });
     const result = await S3.putObject(world.getCommandParameters());
+    const versionId = extractPropertyFromResults<string>(result, 'VersionId');
+    world.saveCreatedObject(world.getSaved<string>('objectName'), versionId || '');
```
`objectName` is specified in parameter, should be the "source of truth" in this function...
(it is too easy to forget to read it from saved and use the variable directly... which may lead to issue when refactoring/...)

```suggestion
async function putObject(world: Zenko, objectName: string = Utils.randomString()) {
    world.addToSaved('objectName', objectName);
    await uploadSetup(world, 'PutObject');
    world.addCommandParameter({ key: objectName });
    world.addCommandParameter({ bucket: world.getSaved<string>('bucketName') });
    const result = await S3.putObject(world.getCommandParameters());
    const versionId = extractPropertyFromResults<string>(result, 'VersionId');
    world.saveCreatedObject(objectName, versionId || '');
```

francoisferrand commented on tests/ctst/steps/pra.ts:
```diff
@@ -298,10 +299,7 @@ When('the DATA_ACCESSOR user tries to perform PutObject on {string} site', { tim
-        this.addCommandParameter({ key: `${Utils.randomString()}` });
-
-        this.setResult(await S3.putObject(this.getCommandParameters()));
+        await putObject(this, Utils.randomString());
```
```suggestion
        await putObject(this);
```

francoisferrand commented on .github/actions/archive-artifacts/action.yaml:
```diff
@@ -42,3 +42,22 @@ runs:
+        MONGODB_COLLECTIONS=$(kubectl exec -n ${NAMESPACE} data-db-mongodb-sharded-mongos-0 -- mongo ${ZENKO_MONGODB_DATABASE} -u ${MONGODB_ROOT_USERNAME} -p ${MONGODB_ROOT_PASSWORD} --authenticationDatabase admin --quiet --eval "db.getCollectionNames().join(' ')")
+
+        for collection in ${MONGODB_COLLECTIONS}; do
+            kubectl exec -n ${NAMESPACE} data-db-mongodb-sharded-mongos-0 -- mongo ${ZENKO_MONGODB_DATABASE} -u ${MONGODB_ROOT_USERNAME} -p ${MONGODB_ROOT_PASSWORD} --authenticationDatabase admin --quiet --eval "db.getCollection((\"${collection}\")).find().toArray()" > /tmp/artifacts/data/${STAGE}/mongodb-${collection}.log
```
can't we use `mongodump` instead?
it allows dumping the whole DB or specific collection, may be faster. However, it exports a BSON: more compact, but maybe less practical to use (for analysis, may need to pass it through `bsondump` or restore it into another mongo)

otherwise, we may use `mongoexport` to dump a collection to JSON, one at a time...

williamlardier replied:
Updated with mongodump & bsondump, let's see if my approach is faster...

williamlardier replied:
New logic takes 10s to complete, I think we are good...

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/2.7/bugfix/ZENKO-4898` with contents from `bugfix/ZENKO-4898`
and `development/2.7`.

Please resolve the conflict on **the integration branch** (`w/2.7/bugfix/ZENKO-4898`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/2.7/bugfix/ZENKO-4898
 $ git pull  # or "git reset --hard origin/w/2.7/bugfix/ZENKO-4898"
 $ git merge origin/development/2.7
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/bugfix/ZENKO-4898
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/bugfix/ZENKO-4898
```





*The following options are set:* **create_integration_branches**

Seragonia committed:
Code formatting

Issue: ZENKO-4898

Seragonia committed:
Unify bucket creation and cleanup

- Remove duplicated logic
- Use a world-managed function to keep track of created objects
- Clean the buckets at the end of any test, unless it failed, to
  ease debugging.

Issue: ZENKO-4898

Seragonia committed:
Cleanup created accounts
- Assume role tests create additional accounts that we must clean
  at the end of any (successful) scenario.
- This ensures that the GetRolesForWebIdentity calls are not
  impacted by the high number of accounts.

Issue: ZENKO-4898

Seragonia committed:
Simplify and unify pradb name

- Quotas are not needed, plus, they cause errors when using the
  env variable in a kubectl command.

Issue: ZENKO-4898

bert-e commented:
# History mismatch



Merge commit #c2e65127c6f4162bc8e06aa73caeb351601707fc on the integration branch
`w/2.7/bugfix/ZENKO-4898` is merging a branch which is neither the current
branch `bugfix/ZENKO-4898` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `bugfix/ZENKO-4898` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

ghost commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.7/bugfix/ZENKO-4898` with contents from `bugfix/ZENKO-4898`
and `development/2.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.7/bugfix/ZENKO-4898 origin/development/2.7
 $ git merge origin/bugfix/ZENKO-4898
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/bugfix/ZENKO-4898
```





*The following options are set:* **create_integration_branches**

Seragonia committed:
Dump mongodb at the end of the tests

- Tests should cleanup resources, unless a scenario failed
- Use of mongodump and bsondump for fast speed

Issue: ZENKO-4898

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-4898` into
`development/2.6`
* [w/2.7/bugfix/ZENKO-4898](https://github.com/scality/zenko/compare/development/2.7...w/2.7/bugfix/ZENKO-4898) will be merged into `development/2.7` 
* [w/2.8/bugfix/ZENKO-4898](https://github.com/scality/zenko/compare/development/2.8...w/2.8/bugfix/ZENKO-4898) will be merged into `development/2.8` 
* [w/2.9/bugfix/ZENKO-4898](https://github.com/scality/zenko/compare/development/2.9...w/2.9/bugfix/ZENKO-4898) will be merged into `development/2.9` 
* [w/2.10/bugfix/ZENKO-4898](https://github.com/scality/zenko/compare/development/2.10...w/2.10/bugfix/ZENKO-4898) will be merged into `development/2.10` 



The following branches will **NOT** be impacted:

* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -31,25 +31,33 @@ export async function cleanS3Bucket(
+        await S3.deleteBucketLifecycle(world.getCommandParameters());
+        await S3.deleteBucket(world.getCommandParameters());
+    } catch (err) {
+        world.logger.warn('Error cleaning bucket', { bucketName, err });
```
whu could this happen?
except a bug in the code, it should not happen IMHO.... so having an error is actually a good thing!

Even if it may not be related to the "current" test, this should not be masked (let's be honest: we will never notice a warning, unless we have another issue)... Esp. if it allows to "detect" unexpected system instability (like high cpu usage or pod restarts), it is better to have an error here and fix the root cause of the instability.

williamlardier replied:
Some tests actually test the deletion of a bucket, or delete a bucket in their logic; in this case and because this can be dynamic, we do not want to fail a successful scenario in this case, because we get a "NoSuchBucket" error. I don't want to filter out this error as well because it may be a real bug in some other scenarios. So having a warning log seem the good trade-off.

francoisferrand replied:
> I don't want to filter out this error as well because it may be a real bug in some other scenarios. So having a warning log seem the good trade-off.

Correct me if I am wrong, but here it seems to me we are filtering all errors : so we will filter-out `NoSuchBucket` errors as well as all the others, and just adding a log (which we will not look at unless we have another test failure)
→ If there is a bug causing NoSuchBucket we will not _notice_
→ If there is a bug causing other error, we will not _notice_ either

> we do not want to fail a successful scenario in this case, because we get a "NoSuchBucket" error

If this was the test, it would fail systematically ; yet the problem we are facing now is flaky: so I don't understand how it relates..

> Some tests actually test the deletion of a bucket, or delete a bucket in their logic

If one test deletes a bucket (or object), it should delete it from the "world" ; or we may add a tag/label on the test, to skip the cleanup when that tag is present...

williamlardier replied:
> Correct me if I am wrong, but here it seems to me we are filtering all errors : so we will filter-out NoSuchBucket errors as well as all the others, and just adding a log (which we will not look at unless we have another test failure)
→ If there is a bug causing NoSuchBucket we will not notice
→ If there is a bug causing other error, we will not notice either

Exactly, I do not want any scenario to fail due to unrelated logic during cleanup: we already have scenarios covering the DeleteBucket API. Throwing errors will not bring anything to our flakiness nor help us detecting bugs for this API, as it should be covered by dedicated scenarios. We can also have buckets where compliance locking is enabled and so we cannot remove the objects. It is totally fine to keep some buckets when we cannot naively clean them.

> If this was the test, it would fail systematically ; yet the problem we are facing now is flaky: so I don't understand how it relates..

I think your point is not about flakiness (it is already solved) nor bug detection (we have dedicated scenarios - the cleanup will fail if these scenarios fail as well), but about enforcing that each scenario complies with the cleanup logic; i.e, we do not leave the "saved" variables in a state where we still have a reference to a bucket that was deleted. But we do not want to delete the bucket from the world after we delete it, because we might need it again (e.g., bucket re-creation). We can add a new flag to say "this bucket was deleted", but it imposes a new logic in our tests (logic that is unknown unless we actually run the scenario and end up with a failure because we tried to delete an already-deleted bucket).
So I think it is better not to alter any existing scenario, and be very flexible. We rely on the dedicated scenarios to detect bugs around bucket cleaning.

francoisferrand replied:
>  fail due to unrelated logic during cleanup: we already have scenarios covering the DeleteBucket API

when I talk about bug, it is not necessarily a bug in the DeleteBucket API (which should be covered explicitely indeed), but more about other bugs... e.g. if the system is unstable or the test itself has an issue, being strict allows to surface these issues.

the point is that we cannot garantee that the tests scenarios will cover any bug : so unless there is a known reason to fail, it should not fail... and it is better to fail (even if hard to detect) than ignore a possible bug (which by definition is not covered by our tests). and leaving the bucket unexpectedly present may still cause more errors later, as we would not be in the "nominal" case of the tests.

> We can add a new flag to say "this bucket was deleted", but it imposes a new logic in our tests (logic that is unknown to the dev unless we actually run the scenario and end up with a failure because we tried to delete an already-deleted bucket)

it is unknown until we tell them it is how it works; and if they don't, they will soon discover it since the test will fail.
the real question here is how many test (or test suites?) would be impacted: do we really have many scenarios which delete their own bucket?

(also I guess deleting a bucket is done is only a few dedicated steps: si it should be handled there, and further limit the changes...)

williamlardier replied:
I am doing it, but:

- I am not aligned for the reasons above, plus ->
  - If the test itself has an issue, it should fail; if not the test should be re-implemented. We shouldn't rely on hooks to detect that.
  - At first, buckets were not cleaned for a conscious reason: we want to have a "living platform" where multiple operations are affecting multiple resources, to be realistic. That's why each scenario today is fully independent from others (thanks to random bucket names, or different accounts when required), excluding lifecycle. If we have a bucket that is failing to be deleted, it does not mean that the scenario is responsible for this error. So we might spend more time trying to understand links between a failure and some transient state of the system, than letting the right scenarios fail. Sometimes we could have an overload in the CI (or anything, due to very valid reasons) and thus make this cleanup fail. In this case, the goal is to still pass the other scenarios, and not fail the current one. The philosophy behind CTST is that we are testing what customers do: instabilities happen, other operations might exhaust the resources... That's fine: as long as the scenario achieves its objectives, then the platform is robust enough. CTST tests feature scenarios, what the customer will use. We should only fail a scenario when its objectives are not achieved. If the instability has no effect on the scenario results, all good. If we want to remove all instabilities sources, CTST is not the right tool...

Also we should really unify our test suites across branches, and not over-engineer our scenarios we have in 2.7+ branches: they re-implement lots of existing logic, do not follow best practices of being independent and parallelizable (such as lifecycle - we already know that), making it very slow to integrate changes targeting all branches, due to both conflicts and errors during test execution, as it is hard to track dependencies between a step and what it "requires" in order to work. E.g. here, when we modify the "saved" resource, we end up breaking tests that re-implemented the logic in a different place...

williamlardier replied:
Done with added commits specific for 2.7+ branch: https://github.com/scality/zenko/compare/development/2.7...w/2.7/bugfix/ZENKO-4898

francoisferrand commented on tests/ctst/steps/utils/utils.ts:
```diff
@@ -196,18 +196,21 @@ async function createBucketWithConfiguration(
+    if (!finalObjectName) {
+        finalObjectName = `${Utils.randomString()}`;
+    }
+    world.addToSaved('objectName', finalObjectName);
```
is it needed? It this will be done by `putObject()` on line 209...

williamlardier replied:
The line 209 only calls the S3 API wrapper from cli-testing, that will either run a cli or sdk command, but it is already outside of the world's scope, and it won't change the 'objName', only consume the input parameters...

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4898 contains:


* `2.10.1`

* `2.6.68`

* `2.7.64`

* `2.8.44`

* `2.9.21`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.10.2`

* `2.6.68`

* `2.7.64`

* `2.8.44`

* `2.9.21`


Please check the `Fix Version/s` of ZENKO-4898, or the target
branch of this pull request.




*The following options are set:* **create_integration_branches**

Seragonia committed:
Throw when bucket cannot be cleaned

- Any bucket with Compliance retention is ignored
- Governance is accepted as we delete the bucket with the account
  identity.

Issue: ZENKO-4898

Seragonia committed:
Handle user metadata and body cleanup

Issue: ZENKO-4898

ghost commented:
/help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **create_integration_branches** | Allow the creation of integration branches. |  | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 






*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

ghost commented:
/after_pull_request=2160

bert-e commented:
# Waiting for other pull request(s)



The current pull request is locked by the **after_pull_request** option.

In order for me to merge this pull request, run the following actions first:


:arrow_right: Merge the `OPEN` pull request:

* pull request #2160





Alternatively, delete all the **after_pull_request** comments from this pull request.





*The following options are set:* **after_pull_request, create_integration_branches**

ghost commented:
/approve

bert-e commented:
# Waiting for other pull request(s)



The current pull request is locked by the **after_pull_request** option.

In order for me to merge this pull request, run the following actions first:


:arrow_right: Merge the `OPEN` pull request:

* pull request #2160





Alternatively, delete all the **after_pull_request** comments from this pull request.





*The following options are set:* **after_pull_request, approve, create_integration_branches**

ghost commented:
ping

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`

* :heavy_check_mark: `development/2.10`



The following branches will **NOT** be impacted:

* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **after_pull_request, approve, create_integration_branches**

bert-e commented:
# Queue build failed




The corresponding build for the queue failed:

- Checkout the [status page](https://bert-e.scality.net/scality/zenko).
- Identify the failing build and review the logs.
- If no issue is found, re-run the build.
- If an issue is identified, checkout the steps below to remove
  the pull request from the queue for further analysis and maybe rebase/merge.

<details>
  <summary><b>Remove the pull request from the queue</b></summary>

- Add a `/wait` comment on this pull request.
- Click on login on the [status page](https://bert-e.scality.net/scality/zenko).
- Go into the [manage](https://bert-e.scality.net/scality/zenko/manage) page.
- Find the option called `Rebuild the queue` and click on it.
  Bert-E will loop again on all pull requests to put the valid ones
  in the queue again, while skipping the one with the `/wait` comment.
- Wait for the new queue to merge, then merge/rebase your pull request
  with the latest changes to then work on a proper fix.
- Once the issue is fixed, delete the `/wait` comment and
  follow the usual process to merge the pull request.

</details>

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`

* :heavy_check_mark: `development/2.10`



The following branches have **NOT** changed:

* `development/2.5`



Please check the status of the associated issue ZENKO-4898.

Goodbye williamlardier.
